pep-9999.rst

```diff
@@ -0,0 +1,214 @@
+PEP: 9999
+Title: Adding Vulnerability Data for the Simple API for Package Indexes
+Author: Dustin Ingram <di@python.org>
+PEP-Delegate: Donald Stufft <donald@stufft.io>
+Status: Draft
+Type: Standards Track
+Topic: Packaging
+Content-Type: text/x-rst
+Created: 28-Dec-2022
+
+
+Abstract
+========
+
+:pep:`691` defined a JSON form for the "Simple Repository API". This allowed
+clients to more easily query the data that was previously only available in
+HTML, as defined in :pep:`503`.
+
+This proposal adds new fields to the JSON form, which allow it to be used in
+place of PyPI's `JSON API <https://warehouse.pypa.io/api-reference/json.html>`__
+to acquire data about known vulnerabilities for a given project.
+
+The new fields are all part of the data returned from the "project details" URL.
+
+The schema for the vulnerability data and related fields is based on the `Open
+Source Vulnerability Schema <https://ossf.github.io/osv-schema/>`__, provided
+by the `Open Source Vulnerability (OSV) <https://osv.dev/>`__ project of the
+`Open Source Software Security Foundation (OpenSSF) <https://openssf.org/>`__.
+
+
+Rationale
+=========
+
+With the introduction of the JSON form of the simple API in :pep:`691`, the
+simple API offers functionality that is almost as complete as the PyPI JSON API.
+This PEP adds data which were previously only available through the JSON API,
+in order to allow more clients which were previously Warehouse specific to
+support arbitrary standards-compliant indexes.
+
+
+Specification
+=============
+
+This specification defines version 1.2 of the simple repository API. For the
+HTML version of the API, there is no change from version 1.0. For the JSON
+version of the API, the following changes are made:
+
+- The ``api-version`` must specify version 1.2 or later.
+- A new ``vulnerabilities`` key is added at the top level.
+- One new "file information" key, ``vulnerability-ids`` is added to the
+  ``files`` data.
+
+The ``vulnerabilties`` and ``vulnerability-ids`` keys are mandatory.
+
+Vulnerabilities
+---------------
+
+An additional key, ``vulnerabilities`` MUST be present at the top level, in
+addition to the keys previous defined in :pep:`691` and :pep:`700`. This key
+MUST contain a mapping from a unique vulnerability identifier to a dictionary
+of fields that describe that vulnerability, for every vulnerability that
+affects the given project.
```

Comment:
Not a serious thing, but how is this expected to interact with PyPI mirrors/mirroring software? Are they expected to mirror PyPI's own JSON API responses (and rewrite the individual file URLs), or should they be prepared to emit a `vulnerabilities: {}` pair if they aren't supplying vulnerability information?

(I suspect a lot of mirrors will stick with the 503-style index/API or just remain at 1.0, so maybe this isn't an issue at all?)

Reply:
Whoops, saw that this is implicitly addressed by https://github.com/di/peps/pull/4/files#diff-93438bd03fb98724e1824ecbf915891c9c401f0128ab7327961ab881c98766aeR128-R129.

Reply:
I think it depends on whether they support PEP 691 as well as this specification (and emit a `"api-version": "1.2"`) or not. If they do, it seems reasonable to mirror the vulnerability data as well if they want. I'll add a FAQ about this.

> (and rewrite the individual file URLs)

All mirrors do this, I'm not sure how it relates though?

Reply:
Makes sense!

That part about rewriting was just me thinking out loud -- normally I'd expect a mirror to download the distributions and generate its own HTML/JSON indices, but for something like a JSON mirror including vulnerability information, it would possibly be easier to download the JSON as well as rewrite the distribution URLs in it. But that's pretty completely irrelevant here 🙂